### ISSUE TITLE
Provide a way to reset a CDC device with unsent flushed data

### ISSUE THREAD

**Issue by dkonigsberg** [hathach/tinyusb#1128]
**Is your feature request related to a problem? Please describe.**
USB CDC does not appear to provide a reliable way of knowing when there is a user application on the host that's actually connected to the device. At best there's the DTR flag, but it cannot always be trusted.  As such, the only way to know if you're transmitting data to nowhere is to check after the fact. (Either by getting a 0 response after successive calls to `tud_cdc_write_flush()` or by waiting on `tud_cdc_tx_complete_cb()` with a timeout.)

Regardless, when you get to this point, data has already been flushed from tinyUSB's internal buffers to the transmit buffer of the device's USB peripheral.  Calling `tud_cdc_write_clear()` will clear out unflushed data, but it won't do anything about flushed data. The flushed data will sit in the device peripheral's own buffer indefinitely, until the user opens a terminal app on the host and connects to the device.

**Describe the solution you'd like**
I would like a way to reset the state of the USB endpoint so that the peripheral's transmit buffer gets cleared. I've done a fair bit of tinkering along these lines, and one possible solution seems to involve explicitly stalling then clearing the endpoint. Specifically, I wrote a function inside `cdc_device.c` that looks like this:

```
uint32_t tud_cdc_n_abort_transfer (uint8_t itf)
{
  cdcd_interface_t* p_cdc = &_cdcd_itf[itf];

  // Skip if usb is not ready yet
  TU_VERIFY( tud_ready(), 0 );

  uint8_t const rhport = TUD_OPT_RHPORT;

  // Claim the endpoint, if not marked busy
  usbd_edpt_claim(rhport, p_cdc->ep_in);

  // Stall the endpoint
  usbd_edpt_stall(rhport, p_cdc->ep_in);

  // Clear the stalled endpoint
  usbd_edpt_clear_stall(rhport, p_cdc->ep_in);

  // Release the endpoint
  usbd_edpt_release(rhport, p_cdc->ep_in);

  return 0;
}
```
It may not be correct, and it may not be the most robust solution, but in my own limited testing it does get the job done.  I'd obviously prefer a more official way of handling this case.

**Comment by hathach** [comment]
@dkonigsberg thanks for the issue, it seems like you want to reset endpoints for CDC. Could you 
1. Explain more about its use case, 
2. what is the issue you are dealing with and how to reproduce it

**Comment by kjrandez** [comment]
I got here from googling when having a similar issue. Basically if you write into the cdc buffer (or vendor ifc buffer in my case) and then the host stops reading it just sits in there. Then when you say open up your PC program again and start reading again it will get that old data. It would be nice to have a way to flush it out on the device side if there is a timeout, but in my case I just made sure to write my host program and communication protocol in a way that at startup it reads out any existing data then resets the system to a known start state -i.e. by writing a reset/start command once there is no more data left to read.
